ethpm_cli/commands/scraper.py

```diff
@@ -75,23 +75,35 @@ def get_ethpm_birth_block(
     Returns the closest block found before the target_timestamp
     """
     version_release_date = datetime.fromtimestamp(target_timestamp)
-    from_date = datetime.fromtimestamp(w3.eth.getBlock(from_block)["timestamp"])
-    delta = version_release_date - from_date
-
-    if delta.days <= 0 and from_date < version_release_date:
-        while (
-            w3.eth.getBlock(from_block)["timestamp"] < version_release_date.timestamp()
-        ):
-            from_block += 1
-        return from_block - 1
 
-    elif from_date < version_release_date:
-        updated_block = int((from_block + to_block) / 2)
-        return get_ethpm_birth_block(w3, updated_block, to_block, target_timestamp)
+    mid_block = int((from_block + to_block) / 2)
 
+    from_date = datetime.fromtimestamp(w3.eth.getBlock(from_block)["timestamp"])
+    to_date = datetime.fromtimestamp(w3.eth.getBlock(to_block)["timestamp"])
+    mid_date = datetime.fromtimestamp(w3.eth.getBlock(mid_block)["timestamp"])
+
+    delta_start = version_release_date - from_date
+    delta_end = version_release_date - to_date
+    delta_mid = version_release_date - mid_date
+
+    min_delta = min(delta_mid.days, delta_start.days, abs(delta_end.days))
```

Comment:
I'm a little confused by what's going on here. Wouldn't the `min_delta` always be `delta_end.days`?

It also seems that adjusting the hunt for the next block by one is going to take a looong time. I'd try looking at [this](https://www.sanfoundry.com/python-program-implement-binary-search-without-recursion/) and see if you can get some better performance by following that pattern.

Reply:
You're right, thanks for that resource. I see now that you have to update the range you're looking at for each iteration